TST: Suppressed warnings

seberg opened the pull request:
OK, this is a start, a context manager to suppress warnings a little better (nothing is prefect, though maybe I am missing something). And trying to remove all those "ignore" stuff everywhere mostly.

A few quite general warnings are still suppressed globally (where possible, sometimes needs local duplication), that may be fine. The switch to only show warnings in release mode is missing, and since some things never showed warnings for me, I sometimes removed it completely. This is not beta or rc, but I think you can get the idea.

matthew-brett commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1961,141 @@ def __exit__(self, *exc_info):
+    of setting a warning to "ignore", it is possible to add a suppression.
+    This has the advantage of ensuring a clean warning registry.
+
+    Suppressions are inserted either immediatly, or upon entering the
```
typo 'immediatly'

seberg commented on numpy/lib/nanfunctions.py:
```diff
@@ -663,12 +663,9 @@ def nanmean(a, axis=None, dtype=None, out=None, keepdims=False):
         raise TypeError("If a is inexact, then out must be inexact")
 
-    # The warning context speeds things up.
-    with warnings.catch_warnings():
```
Didn't understand why it would speed things up, I somewhat assumed that adding divide=ignore for _divide_by_count does the trick here as well. (same below)

seberg replied:
(had to ignore more complex warnings then, but that seemed right)

charris commented:
> The switch to only show warnings in release mode is missing

What do you mean? Currently it depends on the version string.

seberg commented:
Hmmm, python 3.4 and later actually should invalidate the warning registry it appears, so the dance could be unnecessary there.... Not sure what that means.

@charris I mean in this version I broke it currently.

seberg commented:
Ok, this is getting somewhere. The biggest issue for me is that I bet scipy will explode in dev mode at least. I also added stacklevel to most warnings (that I found quickly with a grep), no tests for that sorry, but that would seem insane :).

There are no tests for the new warning context manager yet, but I will probably add them later. Comments?!

seberg commented:
The solution is currently to keep some rather broad global suppressions, but I think that is somewhat fine, since those could be tried to be narrowed down later.

seberg commented:
Hmmm, anyone has an idea why the benchmarks are not successful? Is it an asv problem?!

seberg commented:
OK, I guess adding stacklevel is nonsense. Undid it. I avoided the ASV problem now.

seberg commented:
Hmmm, I readded the stacklevel stuff. I guess we say it is right, I am not quite sure, also it does change things slightly, since the `warnings.filterwarnings(module=...)` will now be the caller instead of us.

seberg commented:
@charris since AppVeyor is so slow, if you would go to teams there and add "GitHub teams", might that give us the option to abort test runs manually?

charris commented:
I can't get anywhere on AppVeyor, their interface is crap. I'll probably need to delete all the cookies and fool around to figure out how to do anything, but IIRC, you need to have an AppVeyor account and then I can add you to the team on AppVeyor. AppVeyor really needs someone who can do interfaces.

seberg commented:
OK, this is actually in a state where it can be reviewed now.

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1964,150 @@ def __exit__(self, *exc_info):
+
+    Note that all of this is only necessary for Python versions before
+    python (about) 3.4. After that "ingnore" can be used safely within
+    the catch_warnings context.
```
Maybe I missed some issue where this was discussed in detail, but I don't really get the point of why this is really needed. Apparently there's a bug in Python 2.7 in the warnings module, which we cannot work around and therefore we need to replace ``catch_warnings`` with a new context manager?

seberg replied:
This is all about the `__warningregistry__` and avoiding getting stuff into it, which catch_warnings does not really help with at all. On newer pythons, this is solved (the registry is invalidated automatically).

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1964,150 @@ def __exit__(self, *exc_info):
+    Context manager doing much the same as ``warnings.catch_warnings``.
+    However, it also provides a suppression mechanism, meaning that instead
+    of setting a warning to "ignore", it is possible to add a suppression.
+    This has the advantage of ensuring a clean warning registry.
```
Since the registry is clean again after leaving the ``with`` block for ``catch_warnings``, does this really matter?

seberg replied:
Wouldn't be surprised if I have to rewrite half of that doc...

The registry is not cleaned again after leaving the with block, this is only possible if we know the exact module, for any other suppression that does not have `module=blah`, this is impossible to do reliably (or would fun file string -> module stuff). This is to just not change it, no matter what we do inside. (again on new pythons probably not a problem due to invalidation of the registry)

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1964,150 @@ def __exit__(self, *exc_info):
+    """
+    Context manager doing much the same as ``warnings.catch_warnings``.
+    However, it also provides a suppression mechanism, meaning that instead
+    of setting a warning to "ignore", it is possible to add a suppression.
```
This sentence basically says ``suppression mechanism == one can add a suppression``, which is fairly self-referential.

seberg replied:
I have updated the text a bit pointing to the python issue, I hope it is somewhat better now. There were also some smaller issues in how the filters worked.

I do not claim this thing is all that great, but I *really* like the end result of this branch, I could just rebase that other new FutureWarning feature on here, and actually find all the tests starting to be flaky immidiatly.

rgommers replied:
It's a lot better now. Still not easy to understand, but that's just because it's not trivial to get your head around how warnings really work under the hood:)

seberg commented:
OK, I have thought about it some more. The new context manager might be mostly not needed. There are two reasons why it is nice (I have to check whether I actually use the second one):

1. Printing the warning is actually possible. Without this context manager, we have to filter out all warnings on the outer level in release mode. So it is either hide everything or raise as an error.

2. It is possible to record more convenient:
```
with suppressed_warnings(record=True) as sup:
    sup.append(FutureWarning, "I can ignore this")
    warnings.filterwarnings("always", category=DeprecationWarning,
                                          message="This is interesting!")
    
    command_that_gives_both_warnings()
    assert_(len(sup.log) == 1)
```
If the futurewarning would not occur every time, counting things would get really annoying. I admit there may be only a single occurance of this (plus the deprecation tests could be made nicer with this, there I needed this kind of behaviour to make the test specific to a single deprecation).

In summery, I still don't care too much about this context manager, but would like to get somewhere with this, because this is just needed to test warnings reliably.
My take is this, if we say we are content with either raising warnings or hiding them and you are unhappy about this rather complicated context manager, then I can get rid of it again and work around it where it is annoying. There is also this `clear_and_catch_warnings` thing that can do the second thing as well with a little more precision needed.

seberg commented:
OK, added a tiny bit more tests, they are not quite perfect coverage, but probably OK. The module based suppression is not too pretty, but should work reliably now (and not different from the rest, except it is more fail-safe from outside influence at the moment). Adding line number based filtering would be simple, but my guess is not needed.

I could add a new feature/different API, and that is a `sup.record`, so that recording unlike catch_warnings, would not record everything but only selective what you want to record. The advantage would be that you can suppress warnings on the outer level and still be selective on the inner one, i.e. I could do that such a thing is possible (assert warns uses its own context):
```
with suppressed_warnings() as sup:
    sup.append(np.ComplexWarning)
    assert_warns(RuntimeWarning, some_function)
```

rgommers commented on numpy/testing/nosetester.py:
```diff
@@ -413,27 +401,62 @@ def test(self, label='fast', verbose=1, extra_argv=None,
                                     module="nose\.")
 
+            # Some of these warnings should likely be more spacific
+            # to a single module, etc.:
```
Indeed, the warnings should only be suppressed when running the numpy tests, not tests of other packages. Otherwise it'll be much harder to fix those other packages up.

seberg replied:
The current state is worse. Would you think it would be OK to leave it for starters (I will look at it, probably some tweaks should be done)?

rgommers replied:
OK no problem.

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1964,155 @@ def __exit__(self, *exc_info):
+    However, it also provides a suppression mechanism to work around
+    http://bugs.python.org/issue4180.
+
+    This bug causes Python before 3.4 to not realiably show warnings again
```
typo ``realiably``

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1964,155 @@ def __exit__(self, *exc_info):
+
+    Paremeters
+    ----------
+    record : boolean
```
nitpick: ``bool`` instead of ``boolean``

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1964,155 @@ def __exit__(self, *exc_info):
+
+    Notes
+    -----
+    Suppressions added inside the the context manager will be discarted
```
typo, should be ``discarded``

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1964,155 @@ def __exit__(self, *exc_info):
+    outer warning handler (i.e. to printing). In recording mode, the
+    `log` attribute will store them.
+
+    If suppressions are added with a module information, the warning
```
"with a module information" --> "and the ``module`` keyword is given"

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1964,155 @@ def __exit__(self, *exc_info):
+
+    If suppressions are added with a module information, the warning
+    registry of this module will additionally be cleared when applying it,
+    entering the context or exiting it.
```
The existing warning registry for the module is cleared upon entering? That doesn't sound like wanted behavior I think. But maybe it's a textual issue in this sentence.

seberg replied:
You are right in some sense, and if you prefer, I could save it and restore it upon exiting (the other warning context does this). On the other hand python3 does this clearing as well. In Py3 "once" is actually "once per change of the warning filters" (this could be considered a bug, but probably a less bad one then the one before).

For testing, I don't think I care to make some "once" warnings appear one time more later on, and when entering, clearing is just being on the safe side.

rgommers commented:
Thanks for the explanations and rewriting the docstring - it's much clearer now why this is needed. +1 from me.

seberg commented:
@rgommers, thanks for the feedback. I think when I have time, I will change the API a little, rename "append" to "filter" probably, and instead of `record=True` make it `sup.record(...)` so that uninteresting warnings are passed to the outer level (maybe also use it for assert_warns). That way we get what I said in the mail with:
```
with suppressed_warnings() as sup:
    sup.filter(UserWarning)
    assert_warns(RuntimeWarning, some_command)
```
not seeing the wrong warning.

seberg commented:
Sorry that it is getting a bit massive :(, I changed the API a bit as I said (maybe the name "suppressed_warnings" is not perfect anymore). I also added a test which checks for both filters with "ignore" and warnings without stacklevel argument using the AST and the example you gave (it is my first dabble with AST, so it is a bit of snipplet pasting in case it looks weird).

seberg commented:
I tested with my ancient scipy, and it runs through fine in "release" mode of course only, in develop mode it cannot obviously, but if I enable my extra test I get:
```
AssertionError: warning filters change during test, added ones:
    "ignore" "None" <class 'scipy.sparse.base.SparseEfficiencyWarning'> "None" 0
    "ignore" "specified build_dir" <type 'exceptions.Warning'> "scipy.weave" 0
    "ignore" "Single-precision types in `eigs` and `eighs`" <type 'exceptions.Warning'> "" 0
    "error" "None" <class 'scipy.signal.filter_design.BadCoefficients'> "None" 0
    "default" "Single-precision types in `eigs` and `eighs`" <type 'exceptions.Warning'> "" 0
    "always" "None" <class 'scipy.signal.filter_design.BadCoefficients'> "None" 0
    "ignore" "None" <class 'numpy.core.numeric.ComplexWarning'> "None" 0
    "always" "None" <class 'scipy.special.basic.SpecialFunctionWarning'> "None" 0
    "ignore" ".*CObject.*" <type 'exceptions.DeprecationWarning'> "bsddb.__init__" 0
```
which makes me wonder.... Some of these like might make sense to get rid of a bit more global/module global, or would it be fine for to use:
```
sup_sparse = suppressed_warnings()
sup_sparse.filter(scipy.sparse.base.SparseEfficiencyWarning)

@sup_sparse.decorate
def test_some_sparse_stuff():
    stuff()
```
on all test funcs, I had a class decorator before, or could look into hacking the run_module_suit to do some stuff.

I am not sure anyway how it plays with scipy, there would currently be missing scipy filters when in dev mode at least, since scipy seems to use numpy's test stuff directly?

seberg commented on numpy/testing/nosetester.py:
```diff
@@ -344,11 +325,17 @@ def test(self, label='fast', verbose=1, extra_argv=None,
+            If True, it will raise an error if the filter list changed during
+            the tests. This should not happen, since all tests should be
+            guarded by a ``catch_warnings`` or similar context manager.
+            If not given, will check if raise_warnings is in "develop" mode.
```
This is an API change, and I am not sure if it is a good idea in any case, though it is probably the easiest way to check if someone got a warnings modification outside a context by accident.

seberg commented on numpy/testing/tests/test_warnings.py:
```diff
@@ -0,0 +1,86 @@
+
+
+import sys
+if sys.version_info >= (3, 4):
```
Not sure if the tokenize is actually necessary to be honest, I stole it from a scipy test (though that one did regex and not AST matching).

seberg commented on numpy/testing/tests/test_warnings.py:
```diff
@@ -0,0 +1,86 @@
+            # default encoding is defined (e.g. LANG='C')
+            with tokenize.open(str(path)) as file:
+                tree = ast.parse(file.read())
+                FindFuncs(path).visit(tree)
```
These tests (outside those few skipped functions), will find calls:

* `warnings.warn(...)` or `warn(...)` without a `stacklevel=` kwarg or 3rd positional argument (same thing)
* `simplefilter` or `filterwarnings` calls with "ignore" as the first positional argument.

rgommers replied:
nice:)

seberg commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1968,231 @@ def __exit__(self, *exc_info):
                 mod.__warningregistry__.update(self._warnreg_copies[mod])
+
+
+class suppressed_warnings(object):
```
Any better name? :)

rgommers replied:
``suppress_warnings`` is already slightly better I think (mirrors ``catch_warnings`` better). Can't think of anything much better than that.

seberg commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1968,231 @@ def __exit__(self, *exc_info):
+
+    Parameters
+    ----------
+    forwarding_rule : str
```
This was mostly to reduce scipy spam, otherwise even a single test (with a loop) might give the same warning lots of times.

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1968,231 @@ def __exit__(self, *exc_info):
+        """
+        Append a new suppression or apply it if the state is entered.
+
+        Paremeters
```
typo ``Parameters``

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1968,231 @@ def __exit__(self, *exc_info):
+
+        Paremeters
+        ----------
+        category : class
```
``, optional``

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1968,231 @@ def __exit__(self, *exc_info):
+        ----------
+        category : class
+            Warning class to filter
+        message : string
```
``str, optional``

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1968,231 @@ def __exit__(self, *exc_info):
+            Warning class to filter
+        message : string
+            Regular expression matching the warning message.
+        module : module
```
``, optional``

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1968,231 @@ def __exit__(self, *exc_info):
+        Append a new recording or apply it if the state is entered.
+        All warnings matching will be appended to the ``log`` attribute.
+
+        Paremeters
```
same typo

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1968,231 @@ def __exit__(self, *exc_info):
+
+    def record(self, category=Warning, message="", module=None):
+        """
+        Append a new recording or apply it if the state is entered.
```
insert blank line

rgommers commented on numpy/testing/utils.py:
```diff
@@ -1937,3 +1968,231 @@ def __exit__(self, *exc_info):
+
+class suppressed_warnings(object):
+    """
+    Context manager doing much the same as ``warnings.catch_warnings``.
```
insert blank line

rgommers commented:
> I am not sure anyway how it plays with scipy, there would currently be missing scipy filters when in dev mode at least, since scipy seems to use numpy's test stuff directly?

I don't think ``scipy`` is special here - ``numpy.testing`` is public and is used by other projects as well. Or I'm missing your point about direct use?

seberg commented:
I meant the `scipy.test()` function. For it some global warning suppressions would be gone. That is fine, but it will force scipy to also clean up more, or wrap it.
Note, all is fine for release mode, at worst it prints quite a few warnings.

rgommers commented:
Regarding ``@sup_sparse.decorate``, I think that would be fine but we'd have to copy the ``suppressed_warnings`` implementation into ``scipy._lib`` (it will be a very long time before we can rely on numpy 1.12.0)

rgommers commented:
> I meant the scipy.test() function. For it some global warning suppressions would be gone. That is fine, but it will force scipy to also clean up more, or wrap it.
Note, all is fine for release mode, at worst it prints quite a few warnings.

Looking at the changes in ``nosetester.py``, I don't see any removed suppressions. Only that in develop mode ``DeprecationWarning, RuntimeWarning`` is changed to ``Warning``, which increases the things that get turned into errors. Which is fine I think.

rgommers commented:
This PR is indeed getting huge. Won't be easy to review in detail. Would it make sense to split off the ``stacklevel`` changes into a separate PR, or are things intertwined (in code or in commit history) too much by now?

seberg commented on numpy/testing/nosetester.py:
```diff
@@ -413,27 +402,55 @@ def test(self, label='fast', verbose=1, extra_argv=None,
-            warnings.filterwarnings("ignore", message="numpy.dtype size changed")
-            warnings.filterwarnings("ignore", message="numpy.ufunc size changed")
-            warnings.filterwarnings("ignore", category=np.ModuleDeprecationWarning)
-            warnings.filterwarnings("ignore", category=FutureWarning)
```
@rgommers, this is what I mean. But I guess it is the only truly removed suppression.

rgommers replied:
IIRC that filter hasn't been present for very long. It looks correct to me to remove it. ``FutureWarning`` should be treated similarly to ``DeprecationWarning`` I'd think.

If this is an issue in develop mode only, I think that's OK. We just have to fix scipy up, preferably right before merging this PR.

seberg replied:
Ah, I was thinking about the boolean negative stuff, but that I did not remove yet (and I think I can just leave it for now). I will see if I manage to test scipy dev with this branch.

seberg replied:
Hah... actually, scipy seems to never runs in development mode, so that the warnings are never set to raise, but only printed out. As such, you might get a print or so more, but actually don't even see any FutureWarning, which is almost surprising.

rgommers replied:
Hmm, looks like you're right. It definitely used to, but after the change to ``raise_warnings='release'``  we didn't update scipy yet. I'll file an issue.

seberg commented:
I think I put them all in separate commits and should be able to split them off, but not tonight ;).

seberg commented:
OK, two changes:

1. I removed the check that there are no accidental additions to the warnings filters. This was mostly because while and a good thing to check once in a while, it was impractical for SciPy, where even one downstream package modifies it.

2. I allowed the feature creap in me to add `log = sup.record(UserWarning)` on top of the `sup.log`, so that you can record the warnings more easily specifically if you want to (just seemed that I have so much machinery to match the warnings correctly, I might as well make it a bit more accessible).

homu commented:
:umbrella: The latest upstream changes (presumably #4619) made this pull request unmergeable. Please resolve the merge conflicts.

seberg commented:
OK, decided to do a bit today. So rebased this. Note that the last commit makes the FutureWarnings with masked arrays show up, so the tests fail of course. Just if anyone still (I care, but only to the point of bringing it up when I happen to cross it) cares about it, my "less noisy" stuff there gives only ~25 failures instead of 65 even without the refcount check (and the refcount check reduces it a lot further).

If anyone finds time to review, I can try to squash it a bit, and will just remove the last commit for the moment. I can also improve the specificity of `assert_warns`, though I think that should wait for a later PR.

seberg commented:
Updated this with the latest necessary warnings suppressions (data attribute, linspace warning, MA warnings), the tests should now pass (should likely rerun test before a possible merge, new things seem to tend to crop up).
Not sure how easy it will be, but probably should try to squash this a little at some point. Right now I am thinking about rather calling it "filter_warnings" instead of "suppress_warnings". But I think it is ready for review (which I probably already had said before ;)).

homu commented:
:umbrella: The latest upstream changes (presumably #7437) made this pull request unmergeable. Please resolve the merge conflicts.

shoyer commented:
Should we also update our testing decorators `assert_warns` and `assert_no_warning` to use `suppress_warnings` rather than `filter_warnings`?

seberg commented:
If we do this, probably, but this is already quite large, so I think I would prefer to do it as a second/third PR.

If you are considering this stuff, the most dubios part is the module filtering. 

On Thu Mar 24 19:25:03 2016 GMT+0100, Stephan Hoyer wrote:
> Should we also update our testing decorators `assert_warns` and `assert_no_warning` to use `suppress_warnings` rather than `filter_warnings`?
> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/numpy/numpy/pull/7099#issuecomment-20095982

seberg commented:
@shoyer, I have had a look, and since it is not a huge change, I included using the new manager for ``assert_warns`` and ``@deprecated`` for now. ``assert_no_warns`` actually does not matter after all, since unless you have a category you are checking for you have to check for any warning, and if you add an "always" filter without category, there is just no difference.
Note that it does change behaviour slightly (to the better, but change). Since this is a testing functionality, I assume that it does not matter much. People who test probably want to know if their tests should be failing ;).

homu commented:
:umbrella: The latest upstream changes (presumably #7421) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #7507) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #7536) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #6872) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #6430) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
Hmm, 34 commits, 900+ new lines. @seberg could you separate out the new context manager into its own PR? The rest is just maintenance and can go in anytime ;)

seberg commented:
Can split it of, I am not sure I understand you though. The maintancance stuff mostly relies on the context manager? Or you just want to start with the context manager I guess?

charris commented:
Right, let's just start with the context manager.

seberg commented:
OK, I will try to rip it apart, probably tomorrow. The commits are likely a mess, but looking over, it can probably be rebuild into new commits almost entirely by splitting along files, so should be easy enough.

charris commented:
Thanks.

seberg committed:
ENH: Make warning testing context managers more specific

This means that warnings of different origin then the one tested for
behave normally. If the normal behaviour is to igonre them this might
decrease specificity in rare cases. In most cases the specificity
will be slightly higher.

seberg committed:
MAINT: Simplify deprecation test decorator

Also modify the corresponding test to suppress the non Deprecation
warnings created to test specificity.

seberg committed:
BUG: Suppress common NaT warnings

Printing of datetime arrays used to cause warning due to
comparison warnings in NaT. This is circumvented by using views
to integer values. Part of this should be simplified when
the deprecation is over.

Also fixes a bug with non-native byteorder.

seberg committed:
MAINT: Add missing teardown in deprecation test case

seberg committed:
ENH: Remove warning ignoring from nanfuncs

Comment mentions a speedup, but it seems unsure why it should
be there. Instead use an error state in divide_by_count.
Some extra complex warnings had to be ignored (but those seemed correct)

seberg committed:
ENH: Use new context manager for testing

Making the outer context manager a suppress warnings gives good
control to print warnings only once in release mode and suppress
some specific warnings which cannot be easily avoided otherwise.

seberg committed:
TST: Use new warnings context manager in all tests

In some places, just remove aparently unnecessary filters.

After this, all cases of ignore filters should be removed from
the tests, making testing (even multiple runs) normally fully
predictable.

homu commented:
:umbrella: The latest upstream changes (presumably #7760) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #7268) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
OK, I rebased and merged #7763, so this needs fixing up.

charris commented:
Can probably clean up the commit history on the way...

seberg commented:
Thanks, I will try to clean this up then in the next few days.

seberg commented:
OK, already got around to doing it, was not too bad anyway and seems like no new updates necessary even. Tried to clean up the git history a bit, but of course the last commit is pretty big....

seberg commented:
OK, will need one more commit, there are some esoteric cases where showing warnings instead of defaulting to error (in other words, switching to release mode), changes things so that warnings are printed in release mode that should not be printed.

There are also two open file handlers python 3 warnings somewhere probably related to the system info tests.

seberg commented:
In case you are wondering, "what the hell?". One example is a case where normally a FutureWarning is raised, and that FutureWarning (if raised) causes a valid return, but also a DeprecationWarning (not great I guess, this is the equality operation, where an error being raised gives a deprecation warning).

So ignoring the FutureWarning would work always (since it is the first warning), but ignoring the DeprecationWarning depends on outer settings.

seberg committed:
ENH: Masked array simplify copying back of unchanged values

Also silences a spurious warning during tests (the multiplication
could give a warning).

seberg committed:
TST: Silence warnings otherwise shown in release mode

These are warnings, which when raised as an error for one reason
or another are already silenced.

seberg commented on numpy/ma/core.py:
```diff
@@ -1032,12 +1032,7 @@ def __call__(self, a, b, *args, **kwargs):
-                masked_da = umath.multiply(m, da)
-                # only add back if it can be cast safely
-                if np.can_cast(masked_da.dtype, result.dtype, casting='safe'):
-                    result += masked_da
```
Not sure what the deal with that block was, I could only guess it did not used to be copyto and whatever was there before did not support broadcasting without trickery. Or maybe I should make it safe casting with the 0?

seberg replied:
The reason for doing this here is that the umath.multiply line causes a RuntimeWarning when setting the warnings to be printed and not raised.

charris replied:
Yeah, that 'safe' casting is a bit weird, as nothing is done if `masked_da` cannot be converted except that the masked values are zeroed rather than set to `masked_da`. It would make more sense if an error or warning was raised on failure to convert.

Hmm, looking at it some more, I think the change here is correct.

seberg committed:
MAINT: Close some filehandlers and PIPEs after being done

The PIPE in the tests caused a ResourceWarning during testing in
python 3.

seberg commented:
OK; managed to track down that last (harmless) resource warning. Added closing of some other file descriptors I ran across trying to figure out what was going on (it was the PIPE's to subprocess.POpen).

charris commented on numpy/core/tests/test_indexing.py:
```diff
@@ -282,8 +282,8 @@ def test_reversed_strides_result_allocation(self):
         # based on ndim instead of size.
-        a = np.full((3, 4, 2), -1)
-        b = np.full((3, 4, 2), -1)
+        a = np.full((3, 4, 2), -1.)
```
Why this change from `1` to `1.`?

seberg replied:
`np.full` issues a FutureWarning because the dtype of the output (was supposed to be same as value, but it was not).

charris replied:
Hmm, I don't see that warning in current master `1.12.0.dev0+a93d9f7`.

seberg replied:
You are right, it was already changed after 1.11 apparently, so the changes are not necessary anymore.

charris commented on numpy/core/arrayprint.py:
```diff
@@ -734,7 +734,9 @@ class TimedeltaFormat(object):
-            v = data[not_equal(data, nat_value)].view('i8')
+            int_dtype = dtype(data.dtype.byteorder + 'i8')
+            int_view = data.view(int_dtype)
+            v = int_view[not_equal(int_view, nat_value.view(int_dtype))]
```
I assume the original was issuing a warning or some such?

charris commented on numpy/core/arrayprint.py:
```diff
@@ -748,7 +750,8 @@ def __init__(self, data):
     def __call__(self, x):
-        if x + 1 == x:
+        # TODO: After NAT == NAT deprecation should be simplified:
+        if (x + 1).view('i8') == x.view('i8'):
```
Needed to avoid warning?

seberg replied:
NAT comparisons give a warning, and they should not while printing in any case. Actually, I am not 100% if you can simplify it currently easily, we may need `np.isnat` or `np.isnan` to support datetimes.

charris commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1976,27 +1976,26 @@ def check_function(self, func, fill_value=None):
-                # do not fill void type
-                if fill_kwarg and dtype.str.startswith('|V'):
-                    continue
+        for size, ndims, order, dtype in itertools.product(*par):
```
No longer a `DeprecationWarning`?

seberg replied:
I did not check everything for unnecessary filters, but apparently yes. Unless it would be platform depenended or so, I would assume its just finished.

charris commented on numpy/core/tests/test_numeric.py:
```diff
@@ -2022,7 +2021,7 @@ def test_for_reference_leak(self):
         np.empty([dim]*10)
         assert_(sys.getrefcount(dim) == beg)
-        np.full([dim]*10, 0)
+        np.full([dim]*10, 0.)
```
Again, why the change in type?

charris commented on numpy/core/tests/test_numeric.py:
```diff
@@ -2199,7 +2198,7 @@ class TestConvolve(TestCase):
         d = [1.] * 100
         k = [1.] * 3
-        assert_array_almost_equal(np.convolve(d, k)[2:-2], np.full(98, 3))
+        assert_array_almost_equal(np.convolve(d, k)[2:-2], np.full(98, 3.))
```
Same question here.

charris commented on numpy/core/tests/test_regression.py:
```diff
@@ -1518,19 +1519,17 @@ def test_ticket_1539(self):
         failures = []
-        # ignore complex warnings
-        with warnings.catch_warnings():
-            warnings.simplefilter('ignore', np.ComplexWarning)
```
No longer a `ComplexWarning`?

seberg replied:
The complex warning only happened on Python 3 because the dtypes dict is unsorted, while on python 2 `a` would always be `np.bool_` and none of the casts need the complex warning then. The cast is not being tested here though.

charris commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -317,26 +317,30 @@ def test_dtype_from_dtype(self):
-                tgt = rf(mat, dtype=np.dtype(c), axis=None).dtype.type
-                res = nf(mat, dtype=np.dtype(c), axis=None).dtype.type
-                assert_(res is tgt)
+                with suppress_warnings() as sup:
```
So this was not needed before. I assume that it is needed now because the lower level warning is no longer caught. Is that what we want?

seberg replied:
Did we used to have a `once` filter maybe, or did we even listen to these warnings at all?

I am not quite sure what you mean here, sorry. I guess the outer level does not catch the warning anymore, you could argue about it, but this is very explicit (i.e. someone writing a new test realizes there is a warning going on).

seberg replied:
Ah, it is `np.nanstd` issuing a warning. Funnily `np.std` does not issue it (I guess this is a slight bug).

charris replied:
Maybe open an issue for `nanstd` with a reference to this line so we can fix it.

charris replied:
I don't think this is the correct fix. The class is a mixin, and fails for one specific function, so suppressing the warning for all functions is over kill. We should either fix the test or fix the function.

seberg replied:
Well, I think emitting the warning is likely the correct thing (and the plain std function is incorrect). Will think about how to make the test more specific.

seberg replied:
Hmmm, yeah I got something wrong. The nanstd seems to be the incorrect version of course, not sure how easy it is to fix (its a rather funny thing, with real input + cast to non-real, but still).

seberg replied:
Hmmm, funny. The `nanstd/nanvar` code uses:
```
np.subtract(arr, avg, out=arr, casting='unsafe')
```
which uses the original arr's dtype, but `avg` already uses the new dtype. I think the spurious complex warning is the only thing that can go really bad though, because only inexact types use this code branch. So the worst that can happen is this warning and a less precise type to store the subtraction result in. If you got a nice idea how to avoid the warning, I am all ears.

seberg replied:
Or maybe I will add a bit non-pretty `if rf in {nanstd, nanvar}: sup.filter(...)` with a comment + new issue....

seberg replied:
I made the test specific with this hack (also fixed the other things like removing the full stuff and used more suppress_warnings, though for now only in the nanfunc tests.... Probably lots of places where we could do it more and I am willing to work a bit more on it later. But for the sake of keeping this not too long, maybe it is better to wait.

seberg replied:
Did not have a quick idea how to fix the spurious warning bug in a very "obvious" way, but can think about it some more....

charris replied:
Yeah, the quick hack was the first thing to occur to me also. Not pretty, but at least it is documented.

charris replied:
> But for the sake of keeping this not too long, maybe it is better to wait.

Yeah, as long as we have a good idea as to what needs fixing, I'd like to get this in as soon as possible.

charris commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -317,26 +317,30 @@ def test_dtype_from_dtype(self):
-                tgt = rf(mat, dtype=c, axis=None).dtype.type
-                res = nf(mat, dtype=c, axis=None).dtype.type
-                assert_(res is tgt)
+                with suppress_warnings() as sup:
```
See previous comment.

charris commented on numpy/lib/tests/test_nanfunctions.py:
```diff
@@ -686,7 +688,7 @@ def test_extended_axis_invalid(self):
     def test_float_special(self):
         with warnings.catch_warnings(record=True):
-            warnings.simplefilter('ignore', RuntimeWarning)
+            warnings.simplefilter('always', RuntimeWarning)
```
Whoa, was that a bug?

charris replied:
So I think ignoring this was intended, what changed?

seberg replied:
Just avoiding all ignore filters, catch_warnings does not propagate outside so setting it to "always" effectively ignores it. Might be a bit cleaner to ues the new context though. At the beginning I did not always.

charris replied:
OK. The record was never looked at anyway, so we could probably do without a filter of anysort. However, using `supress_warnings` might be more informative as that it what is going on.

charris commented on numpy/ma/tests/test_core.py:
```diff
@@ -2289,135 +2302,104 @@ def test_datafriendly_mul_arrays(self):
+            x += t(1)
+            assert_equal(x, y + t(1))
+            xm += t(1)
+            assert_equal(xm, y + t(1))
```
Hmm, the behavior changes here. Before the test failed if any warnings were raised, that is no longer true.

seberg replied:
The outer scope of the tests already ensures that during develop mode with this (unless we change it) and will always print the warning at least once. I could put it back in here, though.

charris commented on numpy/ma/tests/test_core.py:
```diff
@@ -2289,135 +2302,104 @@ def test_datafriendly_mul_arrays(self):
-                assert_equal(xm, y + a)
-                assert_equal(xm.mask, mask_or(m, a.mask))
-
-                assert_equal(len(w), 0, "Failed on type=%s." % t)
```
See previous comment.

charris commented on numpy/ma/tests/test_core.py:
```diff
@@ -2289,135 +2302,104 @@ def test_datafriendly_mul_arrays(self):
+            x -= t(1)
+            assert_equal(x, y - t(1))
+            xm -= t(1)
+            assert_equal(xm, y - t(1))
```
Und so weiter.

seberg replied:
:)

charris commented on numpy/ma/tests/test_core.py:
```diff
@@ -48,6 +49,12 @@
+suppress_copy_mask_on_assignment.filter(
+    numpy.ma.core.MaskedArrayFutureWarning,
+    "setting an item on a masked array which has a shared mask will not copy")
+
```
You know, I have no idea what that `MaskedArrayFutureWarning` means. In fact, I can't even find the message anywhere...

charris replied:
I think it could use rewording.

charris commented on numpy/ma/tests/test_core.py:
```diff
@@ -524,14 +532,14 @@ def test_topython(self):
-        self.assertRaises(TypeError, lambda:float(a))
-        assert_equal(float(a[-1]), 3.)
-        self.assertTrue(np.isnan(float(a[0])))
+            a = array([1, 2, 3], mask=[1, 0, 0])
```
Any particular reason to move these into the context?

seberg replied:
At least the last of these lines also gives the warning, could only put in that one specificily I guess.

charris replied:
So `isnan` is raising a warning?

seberg replied:
the `a[0]` is giving the warning, as it should be (since the element is masked).

charris commented on numpy/ma/tests/test_core.py:
```diff
@@ -1335,24 +1344,28 @@ def test_ne_on_structured(self):
     def test_eq_w_None(self):
```
Be nice to rename this to `test_equal_with_none`.

seberg replied:
Add a renaming commit.

charris commented on numpy/ma/tests/test_extras.py:
```diff
@@ -788,8 +790,7 @@ def test_1d_w_missing(self):
         nx = x.compressed()
         assert_almost_equal(np.cov(nx), cov(x))
         assert_almost_equal(np.cov(nx, rowvar=False), cov(x, rowvar=False))
-        assert_almost_equal(np.cov(nx, rowvar=False, bias=True),
```
Why remove?

seberg replied:
Weird, have to try it. Likely an error because of mistaking it with corrcoef where it wouldn't test much at all since the bias kwarg is deprecated.

seberg replied:
There was no reason, also above the suppression was unnecessary. Reverted it (will squash new commits later).

charris commented on numpy/ma/tests/test_mrecords.py:
```diff
@@ -148,11 +148,9 @@ def test_set_fields(self):
-            rdata['num'] = val
-            assert_equal(rdata.num, val)
-            assert_equal(rdata.num.mask, [1, 0, 0])
+        rdata['num'] = val
```
No more warnings?

seberg replied:
Apparently.... Again, unless this was specific to some platform or so.

seberg replied:
Hmm, may have been that funny future warning, which was more noisy before.

charris commented on numpy/testing/nosetester.py:
```diff
@@ -326,9 +325,9 @@ def test(self, label='fast', verbose=1, extra_argv=None,
+            of 'always' during the test execution.  Valid strings are:
 
-              - "develop" : equals ``(DeprecationWarning, RuntimeWarning)``
+              - "develop" : equals ``(Warning,)``
```
Hmm, this is a bit more severe interpretation of "develop" than before.

charris replied:
So this is why you removed the check for any warnings above? It seems a bit fragile to depend on the setting here. If someone were to change this they would never notice that some tests lost a fraction of their validity.

seberg replied:
Yes, I like the severe interpretation ;). Though scipy might need some fixing up. One good thing is that even if you remove it, the warnings will still get printed once. On the other hand I see no problem with putting back the explicit warning checks in those tests.

seberg replied:
Thought about it some more, and wondering about a middle way, also because it should avoid possible problems with scipy. just leave develop as is, and introduce a new flag `"all"`. Then we could still run some of the Travis tests in "all" mode to ensure that we don't miss odd new warnings/keep the warnings tidy.

charris replied:
Let's leave it as is for the moment. If it causes too much trouble in beta we can always revert.

seberg committed:
MAINT: Spell out with and without in ma test functions

they used to be called `..._w_...` and `..._wo_...`.

charris commented:
I'm a bit concerned that we are now suppressing warnings where that was not previously needed. Was that because of the change in the `develop` mode?

seberg commented:
Well, yeah, though in release mode I still changed it here to print warnings that were not caught explicitly. I do believe we should not silence almost all warnings on the test level (which we did). We can silence some specific ones globally maybe. But would like to notice when a new warnings shows up in lots and lots of the tests.
Of course we could change it to just print the warnings (once), and then basically have a nudge everyone to still strive to get rid of the warnings, but not actually enforce it through failing travis tests (in develop mode).

seberg committed:
TST: Replace catch_warnings when recording is not enforced in test_nanfuncs

seberg committed:
TST: Use ComplexWarning suppression only where needed

seberg commented:
I think I got the last comments. Had some doc fixups and added a note to the release note. Can squash a little when the time comes.

charris commented:
I can put this in as is. If you can easily squash the reversions and some of the fixups, that would be fine also, but don't put too much effort into it.

seberg commented:
OK, got rid of the two REVs (and put that doc thing to the part where its implemented). Should be good.

charris commented:
In it goes, one more step on the way to 1.12. Thanks Sebastian.

seberg commented:
Thanks for the persistent reviewing :).

seberg commented:
Ahhh, just rebased my oindex branch on this, and I get a nice list of all the tests that need modification to be future proof with it :).
